### ISSUE TITLE
Folders for playlists (top panel)

### ISSUE THREAD

**Issue by StrangeSeasOfThought** [Taiko2k/TauonMusicBox#256]
I have and use more playlists now than before. I think it'd be useful to be able to create folders for playlists (top panel):

Main playlist
Playlist 2
[Ratings] -> 4-5 Stars | Unrated
[Released...] -> This year | On this day
[Best of] -> Artist A | Artist B | Artist C
[Genre] -> X | Y | Z
Playlist 3

For compact view, perhaps move all playlists and folders automatically into a single folder:

**{Playlists}**
-> Main playlist
-> Playlist 2
-> [Ratings] -> 4-5 Stars | Unrated
-> [Released...] -> This year | On this day
-> [Best of] -> Artist A | Artist B | Artist C
-> [Genre] -> X | Y | Z
-> Playlist 3

---
Also a themeable color to differentiate smart playlists from normal playlists.

**Comment by farahnur42** [comment]
Piggybacking off of this, not quite related but having a permanent auto-sort option for the playlists would be nice (i.e automatically sorting by a certain order [should ideally be customizable] say `album` -> `artist` -> `track` that persists even after importing new files into the playlist, and auto-sorts said new files according to it)

**Comment by Taiko2k** [comment]
@farahnur42 
One way would be to create two playlists, one for importing, and the other with a [generator code](https://github.com/Taiko2k/TauonMusicBox/wiki/Generator-Codes) applied.

There's also a config file option called `import-auto-sort`, which will apply a `Filepath` > `Album year per artist` to any playlist on any import.

I'm not sure what sorting you exactly have in mind. I assume you meant by `artist` > `album` alphabetically? There's no current way to sort by artist specifically, but by the filepath one should if your music collection is organised.

**Comment by farahnur42** [comment]
Honestly the import auto sort being expanded is what I'm looking for more or less since what I have in mind is just auto-sort after importing/generating according to the tag info @Taiko2k

**Comment by Taiko2k** [comment]
@farahnur42 My current plan now is to remove the existing `import-auto-sort` config option, and replace it a menu option, say, `Make playlist auto sorting` under `Sort...` of the playlist tab.

This new "Make auto sorting" function will just apply a default generator code of `self a path tn ypn auto`. (Codes `self` and `path` are new and not in a release version yet) 

This would, on import, sort by `Folder/Filepath` > `Track numbers` > `Album year per artist`. Did you have a different sorting algorithm in mind / Is there any other sorting function you want not listed in the [generator codes](https://github.com/Taiko2k/TauonMusicBox/wiki/Generator-Codes)?

**Comment by farahnur42** [comment]
@Taiko2k From what I can see that should be fine, although I would personally just prefer alphabetic album sorting instead of alphabetic album sorting by year, that being said it could be made so that the auto-sort option just shows you the generator code tab with the default pre-applied that being what you said and people could modify it accordingly [with maybe a hover/click-over tool-tip either linking to the wiki or just showing you what the main generator codes are]. I think this approach would be more beneficial because it's not anywhere near immediately obvious that the generator codes exist to begin with unless you checked the wiki, which I suspect most people won't do.

**Comment by Taiko2k** [comment]
>  Or could/does just < or > serve as ascending/descending respectively? [i.e tn< meaning ascending track number while tn> meaning descending]

No, those are hard coded, so only the ones listed will work.

>  which I suspect most people won't do.

Yes, in fact only about 1-2% of people who install the app also check the wiki, and even then the portion of people who check any specific page is even less. So yeah probably need to make the generator codes more accessible somehow. Even if its just a hint/link to the page from in the app.

**Comment by farahnur42** [comment]
Does the order matter in these generator codes or will they all apply at the same time? Can't test out right now so that's why I'm asking

**Comment by Taiko2k** [comment]
> Does the order matter in these generator codes or will they all apply at the same time?

They are executed from **left to right**. That can be in what order you choose with the requirement that codes from the *selection* category must come first, then codes from the *search/add* category come second (with at least one required, probably `a`). Then finally choose any combination of *filters* and *sorters*.

**Comment by Taiko2k** [comment]
@StrangeSeasOfThought As for playlist folders... this seems a little too convoluted. The playlist side panel should handle managing playlists good enough.

> Also a themeable color to differentiate smart playlists from normal playlists.

Might be a little over-bearing (an also complicates things since there's already the extra *mouse-over* shade, the *playing* shade and the *selected* shade)

Maybe having an indicator in the playlist list like this is sufficient:

![Screenshot from 2020-03-04 23-43-13](https://user-images.githubusercontent.com/17271572/75871985-43808400-5e72-11ea-90f8-3f4d553c9b8f.png)

**Comment by StrangeSeasOfThought** [comment]
I need the left side panel for the queue... that's why I only use the top panel for playlists.

**Comment by StrangeSeasOfThought** [comment]
> Maybe having an indicator in the playlist list like this is sufficient:

You already have the green square icons, why not simply use the same icon, but in a different color for smart playlists?

I've ended up using symbols:

![symbols1](https://user-images.githubusercontent.com/43582772/76031923-07642680-5f39-11ea-998b-0ae7fd996f00.png)

![symbols](https://user-images.githubusercontent.com/43582772/76031931-0af7ad80-5f39-11ea-9c48-77c67621ee00.png)

**Comment by farahnur42** [comment]
@Taiko2k Speaking of generator codes, could there be ones for the generics i.e album-only, artist-only, duration-only etc? ypa seems to sort some of my songs where I don't want them to be just because they have a certain word in the base album's name [i.e **Monstercat** Uncaged Vol. X]

**Comment by Taiko2k** [comment]
@farahnur42 I'm a little confused. Could you rephrase/elaborate on that?

**Comment by farahnur42** [comment]
I think a couple screenshots would serve to illustrate better, so here they are
What I want: 
![image](https://user-images.githubusercontent.com/37258115/76085795-f53bc400-5fdc-11ea-9990-8a971611ed96.png) [So all songs with the same primary artist involved to be included around the same position)
What is happening: 
![image](https://user-images.githubusercontent.com/37258115/76085646-a8f08400-5fdc-11ea-842c-22413858448c.png) and
![image](https://user-images.githubusercontent.com/37258115/76085923-37fd9c00-5fdd-11ea-9dcf-243136ec4901.png)
I initially assumed this was because tn -> ypa was somehow overriding the artist, so I tried removing ypa from the generator codes, didn't work. So I then tried putting ypa in front of tn but it (the 2nd scenario that is) remained the same. Thankfully sorting by artist after removing ypa from the generator codes does result in the scenario that I want, which is why I think a generator code for each of the main columns by themselves would be a valuable addition.

**Comment by Taiko2k** [comment]
@farahnur42 Yes I could add the column sorts. It had crossed my mind but I hesitated to add it since they are kind of "dumb". But if its producing a desirable output then it could be worth it.

**Comment by farahnur42** [comment]
I think having more generic codes is always more worthwhile in general just coz of the amount of customizability it provides for basically everybody involved, and you can achieve the same end result with more or less variables anyways.

**Comment by Taiko2k** [comment]
@farahnur42 Okay I've added column sorting as a generator code now. For example you can now do `"artist">`

**Comment by farahnur42** [comment]
Nice, time to test it out.

**Comment by Taiko2k** [comment]
@farahnur42 

>  upon further inspection it seems like you missed Album as a column, going by the commit. Adding in "Album" under column_names does solve the issue

Okay I've applied this fix now.
